File: rest/src/main/java/org/example/rest/resources/interactions/EditFollowupMessage.java
package org.example.rest.resources.interactions;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.JsonObject;
import io.vertx.mutiny.uritemplate.Variables;
import org.example.immutables.ImmutableJson;
import org.example.rest.request.Endpoint;
import org.example.rest.request.MultipartRequest;
import org.example.rest.request.Request;
import org.example.rest.request.Requestable;
import org.example.rest.resources.Snowflake;
import org.example.rest.resources.channel.message.AllowedMentions;
import org.example.rest.resources.channel.message.Message;
import org.example.rest.resources.interactions.components.Component;

import java.util.List;
import java.util.Optional;

import static org.example.rest.util.Variables.variables;

@ImmutableJson
@JsonInclude(Include.NON_ABSENT)
public interface EditFollowupMessage extends MultipartRequest, Requestable {

    static Builder builder() {
        return new Builder();
    }

    @JsonIgnore
    Snowflake applicationId();

    @JsonIgnore
    String interactionToken();

    @JsonIgnore
    Snowflake messageId();

    Optional<String> content();

    Optional<List<Message.Embed>> embeds();

    @JsonProperty("allowed_mentions")
    Optional<AllowedMentions> allowedMentions();

    Optional<List<Component>> components();

    Optional<List<Message.Attachment>> attachments();

    @Override
    default Request asRequest() {
        return Request.builder()
                .endpoint(Endpoint.create(HttpMethod.PATCH, "/webhooks/{application.id}/{interaction.token}/messages/{message.id}{?thread_id}", false))
                .variables(variables("application.id", applicationId().getValue(), "interaction.token", interactionToken(), "message.id", messageId().getValue()))
                .body(this)
                .files(files())
                .build();
    }

    class Builder extends ImmutableEditFollowupMessage.Builder {
        protected Builder() {}
    }
}

File: rest/kotlin/src/main/kotlin/json/interaction/EditFollowupMessage.kt
package io.disquark.rest.kotlin.json.interaction

import io.disquark.rest.json.Snowflake
import io.disquark.rest.json.interaction.EditFollowupMessageUni
import io.disquark.rest.kotlin.json.message.EditMessageDsl
import io.disquark.rest.kotlin.nullableoptional.toNullableOptional
import io.disquark.rest.request.Request
import io.disquark.rest.request.Requestable
import io.disquark.rest.request.Requester

class EditFollowupMessage(
    private val requester: Requester<*>,
    private val applicationId: Snowflake,
    private val interactionToken: String,
    private val messageId: Snowflake,
): EditMessageDsl() {
    internal fun toUni(): EditFollowupMessageUni {
        return EditFollowupMessageUni.builder()
            .requester(requester)
            .applicationId(applicationId)
            .interactionToken(interactionToken)
            .messageId(messageId)
            .content(content.toNullableOptional())
            .embeds(embeds?.map { it -> it.map { it.toImmutable() } }.toNullableOptional())
            .allowedMentions(allowedMentions?.map { it.toImmutable() }.toNullableOptional())
            .components(components?.map { it -> it.map { it.toImmutable() } }.toNullableOptional())
            .attachments(attachments?.map { it -> it.map { it.toImmutable() } }.toNullableOptional())
            .build()
    }
}